### PR TITLE
feat: add trun

### PR DESCRIPTION
**Pull request by ryanleecode**
Adds trun cli -- based on pr from capi: https://github.com/paritytech/capi/pull/769

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "[]"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright 2022 Parity Technologies
```
Comment (tjjfvi) [inline LICENSE:190]:
```suggestion
   Copyright 2023 Parity Technologies
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+## Flags
+
+```sh
+--browser -- whether to run tests headlessly in the browser
+--concurrency -- number of tests to be run in parallel
+--dir -- directory to run tests -- relative to cwd
+--filter -- comma delimitated string of files to run -- relative to dir
+--ignore -- name of ignore file -- relative to dir
+--import-map -- name of deno import map file -- relative to cwd
+--output -- name of file to output test results to -- relative to cwd
```
Comment (tjjfvi) [inline README.md:14]:
```suggestion
--browser     # whether to run tests headlessly in the browser
--concurrency # number of tests to be run in parallel
--dir         # directory to run tests -- relative to cwd
--filter      # comma delimitated string of files to run -- relative to dir
--ignore      # name of ignore file -- relative to dir
--import-map  # name of deno import map file -- relative to cwd
--output      # name of file to output test results to -- relative to cwd
```

File: main.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+const controller = new AbortController()
+const { signal } = controller
+
+const ignoreFile = await Deno.readTextFile(path.join(dir, ignore))
+const ignoredFiles = new Set(ignoreFile.split("\n"))
+
+const sourceFileNames = Array.from(Deno.readDirSync(dir))
+  .filter((e) =>
+    e.name.match(/^.*\.ts$/g)
```
Comment (tjjfvi) [inline main.ts:29]:
```suggestion
    /\.ts$/.test(e.name)
```

File: main.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    && e.isFile
+    && !ignoredFiles.has(e.name)
+    && ((filter.size > 0 && filter.has(e.name)) || filter.size === 0)
```
Comment (tjjfvi) [inline main.ts:32]:
```suggestion
    && (!filter.size || filter.has(e.name))
```

File: main.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  self.addEventListener("unload", () => Deno.exit(exitCode))
+  controller.abort()
+}
+
+Deno.addSignalListener("SIGINT", () => shutdown(1))
+Deno.addSignalListener("SIGTERM", () => shutdown(1))
+
+const createBrowser = async () => {
+  const browser = await puppeteer.launch({
+    headless: true,
```
Comment (tjjfvi) [inline main.ts:48]: This should come from a cli flag

File: main.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+  signal.addEventListener("abort", async () => {
+    await browser.close()
+  })
+
+  return browser
+}
+
+const importMapUrl = importMap
+  ? path.toFileUrl(`${Deno.cwd()}/${importMap}`)
```
Comment (tjjfvi) [inline main.ts:61]:
```suggestion
  ? path.toFileUrl(path.resolve(importMap))
```

File: main.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  : undefined
+
+const results: [fileName: string, exitCode: number][] = []
+const runner = browser
+  ? await runWithBrowser({ createBrowser, importMapUrl, results })
+  : await runWithDeno({ reloadUrl: "http://localhost:4646", signal, results })
```
Comment (tjjfvi) [inline main.ts:67]: This reload shouldn't be hardcoded here.

File: deps/pqueue.ts [hunk 8]
```diff
@@ -0,0 +1 @@
+export { default } from "https://deno.land/x/p_queue@1.0.1/mod.ts"
```
Comment (tjjfvi) [inline deps/pqueue.ts:1]:
```suggestion
export { default as PromiseQueue } from "https://deno.land/x/p_queue@1.0.1/mod.ts"
```

File: main.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    "import-map",
+    "output",
+    "reload-url",
+  ],
+  boolean: ["browser", "headless"],
+  default: {
+    "browser-exec-path": "/usr/bin/chromium",
+    headless: true,
+    ignore: ".ignore",
+    "reload-url": "http://localhost:4646",
```
Comment (tjjfvi) [inline main.ts:22]: The default for this shouldn't be capi-specific. I think ideally this should be named `--reload`, with an alias of `-r`, to align with Deno.

File: run.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        outputBuffer.writeSync(new TextEncoder().encode(msg))
+      })
+
+      const exitCode = deferred<number>()
+      await page.exposeFunction("exit", (args: string) => {
+        exitCode.resolve(Number(args))
+      })
+
+      const code = wrapCode(`${consoleJs}\n${buildResult.outputFiles[0]?.text!}`)
+      await page.addScriptTag({ content: code, type: "module" })
```
Comment (tjjfvi) [inline run.ts:63]:
```suggestion
      await page.addScriptTag({ content: consoleJs, type: "module" })
      const code = wrapCode(buildResult.outputFiles[0]?.text!)
      await page.addScriptTag({ content: code, type: "module" })
```

Comment (ryanleecode) [reply run.ts:63]: Not sure if this will work but will give it a try

Comment (ryanleecode) [reply run.ts:63]: Yup seems to be working. B/c its global i guess lol.

**Comment by nythrox** [comment]
Looks good

**Review by nythrox — APPROVED** [review]

File: .vscode/settings.json [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  "[typescript]": {
+    "editor.defaultFormatter": "dprint.dprint"
+  },
+  "deno.codeLens.testArgs": ["-A", "-L=info"],
+  "deno.config": "./deno.jsonc",
+  "deno.suggest.imports.hosts": {
+    "https://deno.land": true,
+    "https://x.nest.land": true,
+    "https://crux.land": true,
+    "http://localhost:4646": false
```
Comment (tjjfvi) [inline .vscode/settings.json:17]:
```suggestion
```

File: main.ts [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+const importMapUrl = importMap
+  ? path.toFileUrl(path.resolve(importMap))
+  : undefined
+
+const results: [fileName: string, exitCode: number][] = []
+const runner = browser
+  ? await runWithBrowser({ createBrowser, importMapUrl, results })
+  : await runWithDeno({ reloadUrl: flags["reload"], signal, results })
```
Comment (tjjfvi) [inline main.ts:81]:
```suggestion
  : await runWithDeno({ reload: flags.reload, signal, results })
```

**Review by tjjfvi — APPROVED** [review]

**Pull request merged**